File: 0079-word-search/0079-word-search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        # Optimization can be made with comparing the count of alphabets in board to that of words. If found less means 
        # a straightforward False
        R=len(board)
        C=len(board[0])
        path=set()
        
        def dfs(r,c,i):
            if i==len(word):
                return True
            if (r<0 or c<0 or r>=R or c>=C or word[i]!=board[r][c] or (r,c) in path):
                return False
            path.add((r,c))
            res = ( dfs(r+1,c,i+1) or
                    dfs(r-1,c,i+1) or
                    dfs(r,c+1,i+1) or
                    dfs(r,c-1,i+1)) 
            path.remove((r,c))
            return res

        for r in range(R):
            for c in range(C):
                if dfs(r,c,0):
                    return True
        return False

File: 0079-word-search/0079-word-search.java
class Solution {
    private char[][] board;
    private int ROWS;
    private int COLS;
    
    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.ROWS = board.length;
        this.COLS = board[0].length;
        
        for (int i = 0; i < this.ROWS; i++){
            for (int j = 0; j < this.COLS; j++){
                if (dfs(i,j,0,word)) return true;
            }
        }
        return false;
    }
    
    private boolean dfs(int row, int col, int ind, String word){
        if (ind >= word.length()) return true;
        
        if ((row < 0 || row >= this.ROWS) || (col < 0 || col >= this.COLS) || (word.charAt(ind) != this.board[row][col])) return false;
        
        boolean ret = false;
        this.board[row][col] = '#';
        
        int[] rowOffset = {0, -1, 0, 1};
        int[] colOffset = {1, 0, -1, 0};
        
        for (int d = 0; d < 4; d++){
            ret = dfs(row + rowOffset[d], col + colOffset[d], ind + 1, word);
            if (ret) break;
        }
        this.board[row][col] = word.charAt(ind);
        return ret;
    }
}